### ISSUE TITLE
Should/could ALE pass --no-fix argument to ruff when linting (to override project configuration)?

### ISSUE THREAD

**Issue by lafrech** [dense-analysis/ale#4756]
Hi. Please bear with my ignorance if I got this wrong.

## Information

**VIM version**

VIM - Vi IMproved 9.0 (2022 Jun 28, compilé May 04 2023 10:24:44)
Rustines incluses : 1-1378, 1499

Operating System: Debian Bookworm

## What went wrong

pyproject.toml file contains the following ruff directives:

```
[tool.ruff]
src = ["src"]
fix = true
show-fixes = true
output-format = "full"
```

I guess the point is that pre-commit fixes everything, which is nice.

The problem is that the due to the `fix` argument, ruff spits out the entire file where ALE expects only json output. ruff ouput parsing fails.

Since ALE distinguishes lint and fix, wouldn't it make sense to force `--no-fix` when linting?

That is, change

    ruff check -q --output-format json-lines --stdin-filename filename - < /tmp/filename

into 

    ruff check -q --no-fix --output-format json-lines --stdin-filename filename - < /tmp/filename

## Reproducing the bug

Add pyproject.toml with

```
[tool.ruff]
fix = true
```

Open a file.

Save file to trigger linters.

### :ALEInfo
<details>
  <summary>Expand</summary>

Current Filetype: python
Available Linters: ['bandit', 'cspell', 'flake8', 'flakehell', 'jedils', 'mypy', 'prospector', 'pycln', 'pycodestyle', 'pydocstyle', 'pyflakes', 'pylama', 'pylint', 'pylsp', 'pyre', 'pyright', 'refurb', 'ruff', 'unimport', 'vulture']
   Linter Aliases:
'jedils' -> ['jedi_language_server']
  Enabled Linters: ['flake8', 'mypy', 'pylint', 'pyright', 'ruff']
  Ignored Linters: []
 Suggested Fixers:
  'add_blank_lines_for_python_control_statements' - Add blank lines before control statements.
  'autoflake' - Fix flake issues with autoflake.
  'autoimport' - Fix import issues with autoimport.
  'autopep8' - Fix PEP8 issues with autopep8.
  'black' - Fix PEP8 issues with black.
  'isort' - Sort Python imports with isort.
  'pycln' - remove unused python import statements
  'pyflyby' - Tidy Python imports with pyflyby.
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'reorder-python-imports' - Sort Python imports with reorder-python-imports.
  'ruff' - A python linter/fixer for Python written in Rust
  'ruff_format' - Fix python files with the ruff formatter.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
  'yapf' - Fix Python files with yapf.
 
 Linter Variables:
" Press Space to read :help for a setting
let g:ale_python_auto_pipenv = 0
let g:ale_python_auto_poetry = 0
let g:ale_python_auto_virtualenv = 0
let g:ale_python_flake8_auto_pipenv = 0
let g:ale_python_flake8_auto_poetry = 0
let g:ale_python_flake8_change_directory = 'project'
let g:ale_python_flake8_executable = 'flake8'
let g:ale_python_flake8_options = ''
let g:ale_python_flake8_use_global = 0
let g:ale_python_mypy_auto_pipenv = 0
let g:ale_python_mypy_auto_poetry = 0
let g:ale_python_mypy_executable = 'mypy'
let g:ale_python_mypy_ignore_invalid_syntax = 0
let g:ale_python_mypy_options = ''
let g:ale_python_mypy_show_notes = 1
let g:ale_python_mypy_use_global = 0
let g:ale_python_pylint_auto_pipenv = 0
let g:ale_python_pylint_auto_poetry = 0
let g:ale_python_pylint_change_directory = 1
let g:ale_python_pylint_executable = 'pylint'
let g:ale_python_pylint_options = ''
let g:ale_python_pylint_use_global = 0
let g:ale_python_pylint_use_msg_id = 0
let g:ale_python_pyright_auto_pipenv = 0
let g:ale_python_pyright_auto_poetry = 0
let g:ale_python_pyright_config = {}
let g:ale_python_pyright_executable = 'pyright-langserver'
let g:ale_python_pyright_use_global = 0
let g:ale_python_ruff_auto_pipenv = 0
let g:ale_python_ruff_auto_poetry = 0
let g:ale_python_ruff_change_directory = 1
let g:ale_python_ruff_executable = 'ruff'
let g:ale_python_ruff_options = ''
let g:ale_python_ruff_use_global = 0
 
 Global Variables:
" Press Space to read :help for a setting
let g:ale_cache_executable_check_failures = v:null
let g:ale_change_sign_column_color = v:null
let g:ale_command_wrapper = ''
let g:ale_completion_delay = v:null
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = v:null
let g:ale_disable_lsp = 'auto'
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%code: %%s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {}
let g:ale_history_enabled = 1
let g:ale_info_default_mode = 'preview'
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = v:null
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'never'
let g:ale_linter_aliases = {}
let g:ale_linters = {}
let g:ale_linters_explicit = 0
let g:ale_linters_ignore = {}
let g:ale_list_vertical = v:null
let g:ale_list_window_size = v:null
let g:ale_loclist_msg_format = v:null
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = v:null
let g:ale_maximum_file_size = v:null
let g:ale_open_list = v:null
let g:ale_pattern_options = v:null
let g:ale_pattern_options_enabled = v:null
let g:ale_root = {}
let g:ale_set_balloons = 1
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = v:null
let g:ale_sign_error = v:null
let g:ale_sign_info = v:null
let g:ale_sign_offset = v:null
let g:ale_sign_style_error = v:null
let g:ale_sign_style_warning = v:null
let g:ale_sign_warning = v:null
let g:ale_sign_highlight_linenrs = v:null
let g:ale_type_map = {}
let g:ale_use_neovim_diagnostics_api = 0
let g:ale_use_global_executables = v:null
let g:ale_virtualtext_cursor = 'all'
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
 
  Command History:

(finished - exit code 127) ['/bin/bash', '-c', '''pylint'' --version']

<<<OUTPUT STARTS>>>
/bin/bash: ligne 1: pylint : commande introuvable
<<<OUTPUT ENDS>>>

(executable check - failure) flake8
(executable check - failure) mypy
(executable check - failure) pylint
(executable check - failure) pyright-langserver
(executable check - success) /home/jerome/.virtualenvs/myproject/bin/ruff
(finished - exit code 0) ['/bin/bash', '-c', 'cd ''/myproject' && ''/home/jerome/.virtualenvs/venv-myproject/bin/ruff'' --version']

<<<OUTPUT STARTS>>>
ruff 0.3.7
<<<OUTPUT ENDS>>>

(finished - exit code 0) ['/bin/bash', '-c', 'cd ''/myproject' && ''/home/jerome/.virtualenvs/venv-myproject/bin/ruff'' check -q --output-format json-lines --stdin-filename ''/myproject/database.py'' - < ''/tmp/vp7Cf6l/7/database.py''']

<<<OUTPUT STARTS>>>
"""Database access"""

import os

"""
---- >8 ----
Cutting actual Python code
---- >8 ----
"""
<<<OUTPUT ENDS>>>

(finished - exit code 127) ['/bin/bash', '-c', '''pylint'' --version']

<<<OUTPUT STARTS>>>
/bin/bash: ligne 1: pylint : commande introuvable
<<<OUTPUT ENDS>>>

(executable check - failure) flake8
(executable check - failure) mypy
(executable check - failure) pylint
(executable check - failure) pyright-langserver
(finished - exit code 0) ['/bin/bash', '-c', 'cd ''/myproject' && ''/home/jerome/.virtualenvs/venv-myproject/bin/ruff'' check -q --output-format json-lines --stdin-filename ''/myproject/database.py'' - < ''/tmp/vp7Cf6l/8/database.py''']

<<<OUTPUT STARTS>>>
"""Database access"""

import os

"""
---- >8 ----
Cutting actual Python code
---- >8 ----
"""

<<<OUTPUT ENDS>>>
</details>

**Comment by lafrech** [comment]
I confirm that adding `--no-fix` fixes it for me.

```vim
    " NOTE: ruff version `0.1.0` deprecates `--format text`
    return ale#Escape(l:executable) . l:exec_args . ' -q'
    \   . ale#Pad(ale#Var(a:buffer, 'python_ruff_options'))
    \   . (ale#semver#GTE(a:version, [0, 1, 0]) ? ' --output-format json-lines' : ' --format json-lines')
    \   . (ale#semver#GTE(a:version, [0, 0, 69]) ? ' --stdin-filename %s -' : ' %s')
    \   . ' --no-fix'
```

This might screw up the fix feature, though.

I'm not familiar with the code base, so I'm leaving this here for anyone to take.

**Comment by yining** [comment]
Hi @lafrech 

Thanks for the bug report. Will submit a PR to fix this, and at the meantime, could you help identify the earliest version of `ruff` that introduced the `--no-fix` flag? I did not recall/know it existed when `ruff` as a linter was initially added to `ALE`.

**Comment by lafrech** [comment]
Thanks @yining .

It is a bit unclear to me when `--no-fix` was introduced.

I've been git-bisecting and the first explicit reference to it that I found is in https://github.com/astral-sh/ruff/commit/ae2ac905dc047c2824d76410f9310e8210623781, which is only a doc update.

Looks like I can't just grep for it. Maybe the argument is treated as the negation of `--fix`, which was added in https://github.com/astral-sh/ruff/commit/f902d25dc71ac77fceec560a3d059272c7e0b889.
